drivers/usb/usbuhci_new/usbuhci.c

```diff
@@ -2117,273 +2111,248 @@ UhciPollNonIsoEndpoint(IN PUHCI_EXTENSION UhciExtension,
                 NextTD,
                 NextTdPA);
 
-    for ( TD = UhciEndpoint->HeadTD; ; TD = TD->NextHcdTD )
+    for (TD = UhciEndpoint->HeadTD; TD != NextTD && TD != NULL; TD = TD->NextHcdTD)
     {
         DPRINT_UHCI("UhciPollNonIsoEndpoint: TD - %p, TD->NextHcdTD - %p\n",
                     TD,
                     TD->NextHcdTD);
 
-        if (TD != NextTD && TD != NULL)
-        {
-            TD->Flags |= UHCI_HCD_TD_FLAG_DONE;
-            InsertTailList(&UhciEndpoint->ListTDs, &TD->TdLink);
+        TD->Flags |= UHCI_HCD_TD_FLAG_DONE;
+        InsertTailList(&UhciEndpoint->ListTDs, &TD->TdLink);
 
-            if (TD->NextHcdTD &&
-                TD->NextHcdTD->HwTD.ControlStatus.Status & UHCI_TD_STS_ACTIVE)
+        if (TD->NextHcdTD &&
+            TD->NextHcdTD->HwTD.ControlStatus.Status & UHCI_TD_STS_ACTIVE)
+        {
+            if (NextTdPA == 0)
             {
-                if (NextTdPA == 0)   
-                    break;
+                TD = TD->NextHcdTD;
+                goto EnqueueTD;
```

Comment:
You skip the check for `NextTd == NULL` on L2155 (that jumps to `ProcessListTDs`), is that intentional?